File: Algorithms/p018_4Sum/p018_4Sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        n = len(nums)
        for i in range(n-3):
            if i>0 and nums[i]==nums[i-1]:
                continue
            for j in range(i+1,n-2):
                if j>i+1 and nums[j]==nums[j-1]:
                    continue
                l = j+1
                r = n-1
                target1 = target-nums[i]-nums[j]
                while l<r:
                    if nums[l]+nums[r]<target1:
                        l += 1
                    elif nums[l]+nums[r]>target1:
                        r -= 1
                    else:
                        res.append([nums[i],nums[j],nums[l],nums[r]])
                        l += 1
                        r -= 1
                        while l<r and nums[l]==nums[l-1]:
                            l += 1
                        while l<r and nums[r]==nums[r+1]:
                            r -= 1
        return res

File: Algorithms/p018_4Sum/p018_4Sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> res;
        int n = nums.size();
        for(int i=0;i<n-3;++i){
            if(i>0&&nums[i]==nums[i-1]) continue;
            for(int j=i+1;j<n-2;++j){
                if(j>i+1&&nums[j]==nums[j-1]) continue;
                int l = j+1;
                int r = n-1;
                int target1 = target-nums[i]-nums[j];
                while(l<r){
                    if(nums[l]+nums[r]<target1){
                        ++l;
                    }else if(nums[l]+nums[r]>target1){
                        --r;
                    }else{
                        res.push_back({nums[i],nums[j],nums[l],nums[r]});
                        ++l;
                        --r;
                        while(l<r&&nums[l]==nums[l-1]) ++l;
                        while(l<r&&nums[r]==nums[r+1]) --r;
                    }
                }
            }
        }
        return res;
    }
};
